src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MtAmazonDynamoDbStreamsBySharedTable.java

```diff
@@ -43,30 +44,108 @@ public GetShardIteratorResult getShardIterator(GetShardIteratorRequest request)
-        do {
-            GetRecordsResult result = dynamoDbStreams.getRecords(request.clone().withShardIterator(iterator));
-            iterator = result.getNextShardIterator();
-            List<Record> records = result.getRecords();
-            if (records.isEmpty()) {
-                break;
-            }
-            records.stream().map(recordMapper).filter(recordFilter).collect(toCollection(() -> mtRecords));
-        } while (mtRecords.isEmpty() && iterator != null);
+        final MtGetRecordsResult result = new MtGetRecordsResult()
+            .withRecords(new ArrayList<>(limit))
+            .withNextShardIterator(request.getShardIterator())
+            .withLastSequenceNumber(null);
+
+        while (addRecords(result, limit, recordMapper, recordFilter)
+            && result.getRecords().size() < limit
+            && result.getNextShardIterator() != null
+            && (mtDynamoDb.getClock().millis() - time) <= timeLimit) {
+        }
 
-        return new GetRecordsResult().withRecords(mtRecords).withNextShardIterator(iterator);
+        return result;
     }
 
+    /**
+     * Fetches more records using the next iterator in the result and adds them to the result records. Returns whether
+     * to keep going or not.
+     */
+    private boolean addRecords(MtGetRecordsResult result,
+                               int limit,
+                               Function<Record, MtRecord> recordMapper,
+                               Predicate<MtRecord> recordFilter) {
+        // retrieve max number of records from underlying stream (and retry below if we got too many)
+        final GetRecordsRequest innerRequest = new GetRecordsRequest().withLimit(MAX_LIMIT)
+            .withShardIterator(result.getNextShardIterator());
+        GetRecordsResult innerResult = dynamoDbStreams.getRecords(innerRequest);
+        result.setNextShardIterator(innerResult.getNextShardIterator());
+        List<Record> records = innerResult.getRecords();
+        // if we did not any records, no point continuing
+        if (records.isEmpty()) {
+            return false;
+        }
+
+        // otherwise, transform records and add those that match the filter
+        final int innerLimit = limit - result.getRecords().size();
+        final List<Record> innerMtRecords = new ArrayList<>(innerLimit);
+        int consumedRecords = addRecords(innerMtRecords, innerLimit, records, recordMapper, recordFilter);
+
+        // If we consumed all records, then we did not exceed the limit, so worth continuing (if we have time)
+        if (consumedRecords == records.size()) {
+            result.getRecords().addAll(innerMtRecords);
+            result.setLastSequenceNumber(getLast(records).getDynamodb().getSequenceNumber());
+            return true;
+        } else {
+            // If we exceeded limit: retry with lower limit (can't just return a subset of records, because the next
```

Comment:
the else block will be reached if consumedRecords < records.size and consumedRecords > records.size

what is expected for this case? `consumedRecords < records.size`, the commend address this case I think `consumedRecords > records.size`

Reply:
An invariant of `addRecords` is that `consumedRecords <= records.size()`, since `consumedRecords` is incremented at most once for each element of `records`. I suppose I could add an assert or something?

Reply:
okay got it